File: Data Structures & Algorithms/partition-equal-subset-sum/submission-1.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int num: nums) {
            sum += num;
        }

        if (sum % 2 != 0) {
            return false;
        }

        Boolean[][] memo = new Boolean[nums.length][sum / 2 + 1];

        return dfs(nums, memo, sum / 2, 0, 0);
    }

    private boolean dfs(int[] nums, Boolean[][] memo, int sum, int curSum, int index) {
        if (curSum == sum) {
            return true;
        }

        if (curSum > sum || index >= nums.length) {
            return false;
        }

        if (memo[index][curSum] != null) {
            return memo[index][curSum];
        }

        memo[index][curSum] = dfs(nums, memo, sum, curSum + nums[index], index + 1)
                            || dfs(nums, memo, sum, curSum, index + 1);

        return memo[index][curSum];

    }
}

File: Data Structures & Algorithms/find-pivot-index/submission-1.py
class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        # idea: trefix sums
        # at an index i, we have two matrices left_sum[] and right_sum[]
        # left_sum[] represents the sum to index i from its left
        # right_sum[] represents the sum from index i to its right
        # we find when left_sum == right_sum, else return -1

        left_sum = [0] * len(nums)
        for i in range(1, len(nums)):
            left_sum[i] = nums[i - 1] + left_sum[i - 1]
        
        right_sum = [0] * len(nums)
        for i in range(len(nums) - 2, -1 , -1):
            right_sum[i] = nums[i + 1] + right_sum[i + 1]
        
        for i in range(len(nums)):
            if left_sum[i] == right_sum[i]:
                return i
        
        return -1